server/lib/chat/chat-service.test.ts

```diff
@@ -0,0 +1,1505 @@
+      asMockedFunction(getUserChannelEntryForUser).mockResolvedValue(null)
+      asMockedFunction(isUserBannedFromChannel).mockResolvedValue(false)
+      asMockedFunction(countBannedIdentifiersForChannel).mockResolvedValue(0)
+      addUserToChannelMock.mockResolvedValue(user1ShieldBatteryChannelEntry)
+      addMessageToChannelMock.mockResolvedValue(joinUser1ShieldBatteryChannelMessage)
+      createChannelMock.mockResolvedValue(shieldBatteryChannel)
+    })
+
+    test("should throw if user doesn't exist", async () => {
+      asMockedFunction(findUserById).mockResolvedValue(undefined)
+
+      await expect(
+        chatService.joinChannel(shieldBatteryChannel.name, user1.id),
+      ).rejects.toThrowErrorMatchingInlineSnapshot(`"User doesn't exist"`)
+    })
+
+    test('should throw if user is banned', async () => {
+      asMockedFunction(isUserBannedFromChannel).mockResolvedValue(true)
+
+      await expect(
+        chatService.joinChannel(shieldBatteryChannel.name, user1.id),
+      ).rejects.toThrowErrorMatchingInlineSnapshot(`"User is banned"`)
+    })
+
+    test('should throw if smurf account is banned', async () => {
+      asMockedFunction(countBannedIdentifiersForChannel).mockResolvedValue(
+        MIN_IDENTIFIER_MATCHES + 1,
+      )
+
+      await expect(
+        chatService.joinChannel(shieldBatteryChannel.name, user1.id),
+      ).rejects.toThrowErrorMatchingInlineSnapshot(`"User is banned"`)
+    })
+
+    test('works when channel already exists', async () => {
+      await chatService.joinChannel(shieldBatteryChannel.name, user1.id)
+
+      expect(addUserToChannelMock).toHaveBeenCalledWith(user1.id, shieldBatteryChannel.id, dbClient)
+      expect(addMessageToChannelMock).toHaveBeenCalledWith(
+        user1.id,
+        shieldBatteryChannel.id,
+        { type: ServerChatMessageType.JoinChannel },
+        dbClient,
+      )
+      expect(client2.publish).toHaveBeenCalledWith(getChannelPath(shieldBatteryChannel.id), {
+        action: 'join2',
+        user: user1,
+        message: {
+          id: joinUser1ShieldBatteryChannelMessage.msgId,
+          type: joinUser1ShieldBatteryChannelMessage.data.type,
+          channelId: joinUser1ShieldBatteryChannelMessage.channelId,
+          userId: joinUser1ShieldBatteryChannelMessage.userId,
+          time: Number(joinUser1ShieldBatteryChannelMessage.sent),
+        },
+      })
+      expect(nydus.subscribeClient).toHaveBeenCalledWith(
+        client1,
+        getChannelPath(shieldBatteryChannel.id),
+        {
+          action: 'init3',
+          channelInfo: shieldBatteryChannel,
+          activeUserIds: [user2.id, user1.id],
+          selfPermissions: channelPermissions,
+        },
+      )
+      expect(nydus.subscribeClient).toHaveBeenCalledWith(
+        client1,
+        getChannelUserPath(shieldBatteryChannel.id, user1.id),
+        undefined,
+      )
+    })
+
+    test("works when channel doesn't exist", async () => {
```

Comment:
nit: avoid `works` when there is some more descriptive phrasing. Here that'd probably be `creates a new channel when it doesn't exist`

Reply:
I also forgot to `expect` that the `createChannel` was actually called in this test, so thanks for that :d